<a href="https://colab.research.google.com/github/itsCiandrei/LinearAlgebra_2ndSem/blob/main/Assignment4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##Linear Algebra for ChE

Matrix Operations

##Laboratory 6: Matrix Operation

###Objectives
At the end of this activity you will be able to:

1.)Be familiar with the fundamental matrix operations.

2.)Apply the operations to solve intemrediate equations.

3.)Apply matrix algebra in engineering solutions.

##Discussion

In [1]:
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline


#Transposition
Transposition is one of the mostly used operations in matrix. Its function is to convert the rows of the matrix into columns. For example, the transpose of the B matrix is: 

$$
B=\begin{bmatrix} 11 & -2 & 95 \\-5 & -11 & 20 \\ 30 & -13 & 73\end{bmatrix}\\
$$

$$
B^T=\begin{bmatrix} 11 & -5 & 30 \\-2 & -11 & -13 \\ 95 & 20 & 73\end{bmatrix}\\
$$

This can also be applied in a program by using the command `np.transpose` or by using the T method

In [2]:
A = np.array([
              [2,3,4],
              [5,3,2],
              [7,9,3]
])
A

array([[2, 3, 4],
       [5, 3, 2],
       [7, 9, 3]])

In [3]:
AE = np.transpose(A)
AE

array([[2, 5, 7],
       [3, 3, 9],
       [4, 2, 3]])

In [4]:
H = A.T
H

array([[2, 5, 7],
       [3, 3, 9],
       [4, 2, 3]])

In [5]:
np.array_equiv(AE,H)

True

In [6]:
H.T.shape

(3, 3)

In [7]:
P = np. array ([
                [9,2,3,11],
                [21,3,6,7],
                [3,4,5,0]
])
JP = np.transpose(P)
JP

array([[ 9, 21,  3],
       [ 2,  3,  4],
       [ 3,  6,  5],
       [11,  7,  0]])

In [8]:
PJ = P.T
PJ

array([[ 9, 21,  3],
       [ 2,  3,  4],
       [ 3,  6,  5],
       [11,  7,  0]])

In [9]:
np.array_equiv(JP,PJ)

True

In [10]:
P.T.shape

(4, 3)

$$
A = 
    \begin{bmatrix}
        x & y \\ 
        4x & 10y
    \end{bmatrix}
 $$


$$
B=\begin{bmatrix} 1 & 2 & 5 \\5 & -1 & 0 \\ 0 & -3 & 3\end{bmatrix}\\
$$

##Dot Product

If you recall the dot product from laboratory activity before, we will try to implement the same operation with matrices. In matrix dot product we are going to get the sum of products of the vectors by row-column pairs. So if we have two matrices $X$ and $Y$:

$$X = \begin{bmatrix}x_{(0,0)}&x_{(0,1)}\\ x_{(1,0)}&x_{(1,1)}\end{bmatrix}, Y = \begin{bmatrix}y_{(0,0)}&y_{(0,1)}\\ y_{(1,0)}&y_{(1,1)}\end{bmatrix}$$

The dot product will then be computed as:
$$X \cdot Y= \begin{bmatrix} x_{(0,0)}*y_{(0,0)} + x_{(0,1)}*y_{(1,0)} & x_{(0,0)}*y_{(0,1)} + x_{(0,1)}*y_{(1,1)} \\  x_{(1,0)}*y_{(0,0)} + x_{(1,1)}*y_{(1,0)} & x_{(1,0)}*y_{(0,1)} + x_{(1,1)}*y_{(1,1)}
\end{bmatrix}$$


Dot product of matrices is where we multiply the rows of the fist matrix to the columns of the second matrix. The formula goes as follows for matrix A and B.

$$
A=\begin{bmatrix} 1 & 2 \\5 & -1 \end{bmatrix}\\
$$

$$
B=\begin{bmatrix} 2 & 6 \\3 & -4 \end{bmatrix}\\
$$

The dot product of A and B is equal to:

$$
AB=\begin{bmatrix} (1)(2) + (2)(3) & (1)(6) + (2)(-4)\\ (5)(2) + (-1)(3) & (5)(6) + (-1)(-4)\end{bmatrix}\\
$$

$$
AB=\begin{bmatrix} 8 & -2\\ 7 & 34\end{bmatrix}\\
$$

The solution can be used in a python program by using `np.dot()`, `np.matmul()`, and `@` operator.


In [12]:
A = np.array([
              [1,2],
              [5,-1]
])
B = np.array([
              [2,6],
              [3,-4]
])

In [13]:
A @ B

array([[ 8, -2],
       [ 7, 34]])

In [14]:
A.dot (B)

array([[ 8, -2],
       [ 7, 34]])

In [15]:
np.matmul(A,B)

array([[ 8, -2],
       [ 7, 34]])

In [16]:
np.dot(A,B)

array([[ 8, -2],
       [ 7, 34]])

#####In matrix dot products there are additional rules compared with vector dot products. Since vector dot products were just in one dimension there are less restrictions. Since now we are dealing with Rank 2 vectors we need to consider some rules:


#Rule 1: The inner dimensions of the two matrix must be equal.
We determine the dimension of the product by checking if it is possible. For example, a (A x B) matrix multiplied by a (B x A) the product matrix will have a (A x A) matrix. This happens because the first dimension of the first matrix and the last dimension of the second matrix determines the dimension of the product if and only if they have the same inside dimension. For example:


$$
RJ=\begin{bmatrix} 9 & -8 & 3 \\76 & -23 & -12 \\ 3 & 1 & 21\\1 & 1 & 1\end{bmatrix}\\
$$

$$
B=\begin{bmatrix} 5 & -8 & -93 \\-16 & -2 & -2 \\ 23 & -21 & 1\end{bmatrix}\\
$$
$$
T=\begin{bmatrix} 1 & -2 & 22\\6 & -3 & 20\end{bmatrix}\\
$$

The P matrix has a dimension of (4 x 3) and the B matrix has a dimension of (3 x 3) and lastly, the T matrix has a dimension of (4 x 3). The only possible dot product would be $T \cdot B$ and $P \cdot B$   




In [17]:
RJ = np.array([
               [9,-8,3],
               [76,-23,-12],
               [3,1,21],
               [1,1,1]
])
CD = np.array([
               [5,-8,-93],
               [-16,-2,-2],
               [23,-21,1],
         
])
LA = np.array([
               [1,-2,22],
               [6,-3,-20],
              
])
print(RJ.shape)
print(CD.shape)
print(LA.shape)

In [ ]:
RJ @ CD

In [ ]:
LA @ CD

In [ ]:
RJ @ LA.T

In [ ]:
CD @ RJ.T

###Rule 2: Dot Product has special properties

#####In the field of data science domain, linear algebra is one of the essential topics, which includes the basic operations in solving matrices: Dot product and matrix multiplication. The Dot product of two matrices is the sum of its elements' products relative to its position; column and rows. 

This basic operation has its own properties, and it should take into consideration when formulating a particular solution:
1. $A \cdot B \neq B \cdot A$ - Noncommutative Property of Multiplication of matrices
2. $A \cdot (B \cdot C) = (A \cdot B) \cdot C$ - Associative Property of Multiplcation of martices
3. $A\cdot(B+C) = A\cdot B + A\cdot C$ - Distributive Property of Addition of matrices
4. $(B+C)\cdot A = B\cdot A + C\cdot A$ - Distributive Property of Addition of matrices
5. $A\cdot I = A$ - Multiplicative Identity Property
6. $A\cdot \emptyset = \emptyset$ - Multiplicative Property of Zero


In [ ]:
a = np.array([
              [1,0,-6],
              [3,27,81],
              [4,25,-36]
])
b = np.array([
              [-21,20,-26],
              [13,7,24],
              [5,6,-6]
])
c = np.array([
              [-3,11,-13],
              [31,17,-2],
              [15,-6,-98]
])

In [ ]:
A1 = a@b
A2 = b@c
B1 = b + c

In [ ]:
##1st Property
np.array_equiv(a@b, b@a)

False

In [ ]:
a@b 

array([[-51, -16,  10],
       [693, 735,  84],
       [ 61,  39, 712]])

In [ ]:
b@a

array([[ -65, -110, 2682],
       [ 130,  789, -375],
       [  -1,   12,  672]])

In [ ]:
##2nd Property
np.array_equiv(a@A2, A1@c)

True

In [ ]:
a@A2

array([[  -193,   -893,   -285],
       [ 21966,  19614, -18711],
       [ 11706,  -2938, -70647]])

In [ ]:
A1@c

array([[  -193,   -893,   -285],
       [ 21966,  19614, -18711],
       [ 11706,  -2938, -70647]])

In [ ]:
##3rd Property
np.array_equiv(a@B1, a@b+a@c)

True

In [ ]:
a@B1

array([[ -144,    31,   585],
       [ 2736,   741, -7947],
       [  284,   724,  4138]])

In [ ]:
a@b+a@c

array([[ -144,    31,   585],
       [ 2736,   741, -7947],
       [  284,   724,  4138]])

In [ ]:
##4th Property
np.array_equiv(B1@a, b@a+c@a)

True

In [ ]:
B1@a

array([[  -87,  -138,  4059],
       [  204,  1198,   888],
       [ -396, -2600,  3624]])

In [ ]:
b@a+c@a

array([[  -87,  -138,  4059],
       [  204,  1198,   888],
       [ -396, -2600,  3624]])

In [ ]:
##5th Property
np.array_equiv(a@np.eye(3,3), a)

True

In [ ]:
a@np.eye(3,3)

array([[  1.,   0.,  -6.],
       [  3.,  27.,  81.],
       [  4.,  25., -36.]])

In [ ]:
a

array([[  1,   0,  -6],
       [  3,  27,  81],
       [  4,  25, -36]])

In [ ]:
##6th Property
np.array_equiv(a@np.zeros(a.shape), np.zeros(a.shape))

In [ ]:
a@np.zeros(a.shape)

array([[0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.]])

In [ ]:
np.zeros(a.shape)

array([[0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.]])

###Determinant


In linear algebra, determinants are mainly used to solve a series of linear equations by applying the concept of Cramer's rule to a set of non-homogeneous equations. To determine the determinants of a certain matrix, one of the requirements is that the sample matrices should be a square matrix. Moreover, there are three types of determinants, first is the singular determinant wherein the value of the determinant of a matrix is zero, and if it is one, then this type of determinant is unimodular. However, for the last type of determinant, in order for a system to have a unique solution, the value of the determinant must be nonzero.

The determinant of some matrix $A$ is denoted as $det(A)$ or $|A|$. So let's say $A$ is represented as:
$$A = \begin{bmatrix}a_{(0,0)}&a_{(0,1)}\\a_{(1,0)}&a_{(1,1)}\end{bmatrix}$$
We can compute for the determinant as:
$$|A| = a_{(0,0)}*a_{(1,1)} - a_{(1,0)}*a_{(0,1)}$$
So if we have $A$ as:
$$A = \begin{bmatrix}2&3\\7&9\end{bmatrix}, |A| = -3$$

For this experiment, we will utilize Python to solve for the determinants of different types of matrices, especially having a dimension greater than 2x2.


In [ ]:
e = ([
      [2,3],
      [7,9]
     
])
np.linalg.det(e)

-3.0000000000000018

In [ ]:
d = ([
      [0,10,1],
      [2,-3,11],
      [-34,-35,21]
])
np.linalg.det(d)

-4332.000000000005

In [ ]:
RR = ([
       [2,4,6,8,10],
       [1,4,9,16,25],
       [-3,5,-7,9,-11],
       [21,12,0,-11,11],
       [1,2,3,4,0]
])
np.linalg.det(RR)

28399.999999999993

###Inverse

The inverse of a square matrix A, also known as reciprocal matrix, is represented as $A^{-1}$. It is one of the crucial tools in mathematics since it is used to find the solution of linear equations through the inversion method. Also, it is frequently used in various fields of science, such as in quantum mechanics and optics.
Now to determine the inverse of a matrix we need to perform several steps. So let's say we have a matrix $M$:
$$M = \begin{bmatrix}-33&12\\6&66\end{bmatrix}$$
First, we need to get the determinant of $M$.
$$|M| = (-33)(66)-(6)(12) = -2250$$
Next, we need to reform the matrix into the inverse form:
$$M^{-1} = \frac{1}{|M|} \begin{bmatrix} m_{(1,1)} & -m_{(0,1)} \\ -m_{(1,0)} & m_{(0,0)}\end{bmatrix}$$
So that will be:
$$M^{-1} = \frac{-1}{2250} \begin{bmatrix} 66 & -12 \\ -6 & -33\end{bmatrix} = \begin{bmatrix} \frac{-11}{375} & \frac{2}{375} \\ \frac{1}{375} & \frac{13}{1125}\end{bmatrix}$$
For higher-dimension matrices you might need to use co-factors, minors, adjugates, and other reduction techinques. To solve this programmatially we can use `np.linalg.inv()`.

In [ ]:
r = np.array([
              [-33,12],
              [6,66]
])
np.array(r@np.linalg.inv(r), dtype=int)

array([[1, 0],
       [0, 1]])

In [ ]:
x = ([
      [10,20,12],
      [42,-31,31],
      [-14,-15,21]
])
y = np.linalg.inv(x)
y

array([[ 0.00454235,  0.01465273, -0.02422585],
       [ 0.03213832, -0.00923122, -0.00473772],
       [ 0.02598418,  0.00317476,  0.0280844 ]])

In [ ]:
x @ y, 

(array([[ 1.00000000e+00, -1.56125113e-17, -5.55111512e-17],
        [-8.32667268e-17,  1.00000000e+00, -1.24900090e-16],
        [ 8.32667268e-17,  1.95156391e-17,  1.00000000e+00]]),)

In [ ]:
X1 = ([
       [21,94,6,-8,0],
       [13,-14,9,-6,2],
       [-23,51,-2,1,-71],
       [2,1,0,-1,1],
       [-1,23,-33,42,10]
])
X2 = np.linalg.inv(X1)
X2

array([[-2.33648239e-03,  5.22520035e-02,  7.22639686e-03,
         2.74692490e-01,  1.33877680e-02],
       [ 1.05871447e-02, -1.03544169e-02, -1.23069641e-03,
        -5.84265231e-02, -8.24408812e-04],
       [ 2.72475659e-02,  7.10082901e-02, -1.11157131e-02,
        -8.78138703e-01, -5.30935053e-03],
       [ 1.37013589e-02,  6.87533282e-02, -3.82817585e-03,
        -6.24047887e-01,  2.14740745e-02],
       [ 7.78717893e-03, -2.53962619e-02, -1.70502732e-02,
        -1.15006344e-01, -4.47705264e-03]])

In [ ]:
np.array(X1@X2, dtype=int)

array([[1, 0, 0, 0, 0],
       [0, 1, 0, 0, 0],
       [0, 0, 1, 0, 0],
       [0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0]])